### DISCUSSION TITLE
Performance improvements / feedback?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#529] [Show and tell]
I recently [pushed](https://github.com/zxing-cpp/zxing-cpp/commit/5a08f7a9cd0095b7f7e37954e03ad9191e31fa62) a change that has significant performance benefits but not across all hardware. I observed 20-30% on my Core i9 (AVX2) but no benefit or even a 50% regression on my Google Pixel 3. Therefore I added the `#ifdef __AVX__` guards.

If anyone wants to experiment with those changes (remove the `#ifdef` lines), please let me know about your results. I'd be interested to know on which architectures this is helpful, neutral or hurtful.

Another change with potential performance implications was [this](https://github.com/zxing-cpp/zxing-cpp/commit/dcfba650a4035f9daf428c268bab0a29cccb0c20). There you have another option to play with.

@markusfisch, @parallaxe, @vkrause, @dkonecny-oracle

EDIT: turns out, my performance measurements on my phone were off, partly because the impact of the phone's temperature on the performance were more pronounced than I thought. Bottom line: I enabled the first change also on non-AVX systems (namely all 64-bit systems) [here](https://github.com/zxing-cpp/zxing-cpp/commit/d2ed8152aa866714348ffb712672d38b64f030b7).

**Comment by markusfisch** [comment]
Ah, the edit clears it up 😉 I tried to profile this yesterday on a couple of devices (Pixel 2, Pixel 3a, Fairphone 3 and 4) and couldn't see any performance regression.

At least for these devices, the measurements varied more between the individual measurement runs than the difference between the old and the new code.

**Reply by axxel** [reply]
Interesting. To clarify: you tested including the change I mentioned in my "EDIT"? That you did not see any noticeable change is surprising in so far as I did in-fact see an improvement on my Pixel 3. It is generally most noticeable with both `tryHarder` and `tryRotate` enabled. Could it maybe also be that some/all of those phones are actually 32bit architectures?

Anyway, the info that I did not make the performance worse is already a very useful feedback. Thanks.

**Reply by markusfisch** [reply]
> you tested including the change I mentioned in my "EDIT"? 

No, I removed the `#ifdef` lines because I didn't notice the new commit yesterday. But I just run a couple of measurements on a Pixel 2 with this commit, just for a quick check, and, at least for this phone, it still makes no real difference. Not even with `tryHarder` and `tryRotate` enabled (which takes a lot longer, of course, but is still similar to the old code).

I am taking the average time for 500 frames for this [statement](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/android/zxingcpp/src/main/java/com/zxingcpp/BarcodeReader.kt#L76) before line 76 and after line 92.

For the Pixel 2, I see around 15ms without `tryHarder` and `tryRotate` and around 50ms with both enabled, for the old and the new code 🤔 I will run more tests later today, because I'm too wondering why I don't see any noticeable difference.

> Could it maybe also be that some/all of those phones are actually 32bit architectures?

No, these phones all have some generation of the [Kryo](https://en.wikipedia.org/wiki/Kryo) architecture which implements the ARM 64-bit instruction set.

**Comment by vkrause** [comment]
Finally got around to run our test suite on 2.0.0, master and master with the binarizer change for comparison. Between 2.0.0 and master I see maybe a 5% speedup (but that's close to the measurement noise), on an i7. That seems in line with your numbers, given that test includes the entire PDF processing pipeline, not just barcode decoding.

Enabling the binarizer change seems to have no impact. This is all with "pure" barcodes though.

**Reply by axxel** [reply]
Thanks for the feedback. Working with "pure" input, you should use the `FixedThreshold` or `BoolCast` binarizer (which you do, iirc). In that case neither the `GlobalHistogramBinarizer` nor the `HybridBinazier` changes would have any impact, indeed.
